### PR TITLE
PCHR-2455: Filter documents by expiry and due date on documents dashboard

### PR DESCRIPTION
**Pull request by bmnepali**
## Overview
This PR reuses the `filterByDue` filter (which was used to filter the documents based on document due date), to filter the documents based on the documents expiry date as well.

## Before
- There was no functionality to filter the documents based on document expiry date.
(no screenshot)

## After
- Now the documents can be filtered based on expiry date along with due date
2455-final-iloveimg-compressed

## Technical Details
1. Renamed the filter name to filterByDueExpiry
```js
filters.filter('filterByDueExpiry', ['$filter', '$log', function ($filter, $log) {
  ...
}])
```

2. Use `field` in the filter to determine which attribute (due or expiry date)  to use for filtering documents.
```js
filters.filter('filterByDueExpiry', ['$filter', '$log', function ($filter, $log) {
  ...
  return function (inputArr, type, field, dateRange) {
     ...
     // use field to filter the document based on due date or expiry date
     if (inputArr[i][field]) {
       itemDueDate = new Date(inputArr[i][field]).setHours(0, 0, 0, 0);
    }
    ...
  }
  ...
}])
```
3. Add a function which will filter the documents based on filter type and by due and expiry date in "controllers/document-list.js"
```js
/**
   * Filters the document list based on filter type and due or expiry date
   * @param {string} type filter type
   * @return {array} Document list
   */
$scope.filterByDateField = function (type) {
    var listByDueDate = $filter('filterByDateField')($scope.list, type, 'activity_date_time', $scope.filterParams.dateRange);
    var listByExpiryDate = $filter('filterByDateField')($scope.list, type, 'expire_date', $scope.filterParams.dateRange);

     return _.uniq(_.union(listByDueDate, listByExpiryDate), 'id');
};
```

4. Add a function which will filter the tasks based on filter type and by due and expiry date in "controllers/task-list.js"
```js
$scope.filterByDateField = function (type) {
    return $filter('filterByDateField')($scope.list, type, 'activity_date_time', $scope.filterParams.dateRange);
};
```
5. Amend the filter to count the document types numbers in file 'views/dashboard/documents.html'
For my documents count:
```html
({{
   ((filterByDateField(filterParams.due)) |
      filterByStatus:filterParams.documentStatus |
      filterByOwnership:'assigned').length
}})
```
For delegated documents count:
```html
({{
   ((filterByDateField(filterParams.due)) |
    filterByStatus:filterParams.documentStatus |
    filterByOwnership:'delegated').length
}})
```
For all documents count:
```html
({{
   (list |
    filterByDateField:filterParams.due:filterParams.dateRange |
    filterByStatus:filterParams.documentStatus).length
}})
```
6. Filter documents based on due date or expiry date in file 'views/dashboard/documents.html'
```html
<tr ng-repeat="document in
      listPaginated = (listFiltered = (listOngoing = (filterByDateField(filterParams.due))
      | filterByOwnership:filterParams.ownership
      | filterByAssignmentType:filterParams.assignmentType
      | filterByContactId:filterParams.contactId
      | filterByStatus:filterParamsHolder.documentStatus
      | orderBy: 'activity_date_time')
      | offset: (pagination.currentPage-1)*pagination.itemsPerPage
      | limitTo: pagination.itemsPerPage)"
    ng-controller="DocumentCtrl"
    ng-include
    src="'document.html'">
</tr>
```
7. Similarly, apply the updated filter in places where `filterByDue` was used in task list, document list. Also in filtering document and tasks in html markup in respective files.

## Comments
- Added additional test cases for the filter 
- Pre-commit issues were also fixed in this PR for ` /controllers/task-list.js` and ` tests/controllers/task-list.spec.js`

- [x] Tests Pass

### CODE REVIEW

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/filters/due-expiry.filter.js [hunk 1]
```diff
@@ -1,93 +1,90 @@
...
 define([
   'tasks-assignments/filters/filters'
 ], function (filters) {
   'use strict';
 
-  filters.filter('filterByDue',['$filter', '$log', function($filter, $log){
-    $log.debug('Filter: filterByDue');
-
-    return function(inputArr, type, dateRange) {
+  filters.filter('filterByDueExpiry', ['$filter', '$log', function ($filter, $log) {
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/filters/due-expiry.filter.js]: The name `filterDueExpiry` is tightly coupled with field names. What if we need to add another field in future, then we would need to change the filter name to `filterDueExpiryXY`etc. Its better to name it something like `filterByDateField`.

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/filters/due-expiry.filter.js [hunk 2]
```diff
@@ -1,93 +1,90 @@
...
-        inputArrlen = inputArr.length,
-        date = new Date(),
-        today = date.setHours(0, 0, 0, 0);
+      var date = new Date();
+      var filteredArr = [];
+      var i = 0;
+      var inputArrlen = inputArr.length;
+      var today = date.setHours(0, 0, 0, 0);
+
+      field = field || 'activity_date_time';
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/filters/due-expiry.filter.js]: You are already sending the field name from the html, then why do we need to default a field name. Instead throw an error, if no field is specified.

File: uk.co.compucorp.civicrm.tasksassignments/views/dashboard/documents.html [hunk 3]
```diff
@@ -40,10 +40,14 @@ <h3 class="panel-title">{{label[filterParams.due]}}</h3>
                 ng-model="filterParams.ownership"
                 uib-btn-radio="'assigned'">
                 My Docs
-              <span>({{(list |
-                filterByDue:filterParams.due:filterParams.dateRange |
-                filterByStatus:filterParams.documentStatus |
-                filterByOwnership:'assigned').length}})
+              <span>
+                ({{
+                  ((((list | filterByDueExpiry:filterParams.due:'activity_date_time':filterParams.dateRange)).length ?
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/views/dashboard/documents.html]: Having this big conditions in the HTMl is hard to follow, its better to create scope function and use the filters in the JS itself.

File: uk.co.compucorp.civicrm.tasksassignments/views/dashboard/documents.html [hunk 4]
```diff
@@ -52,10 +56,14 @@ <h3 class="panel-title">{{label[filterParams.due]}}</h3>
                 ng-model="filterParams.ownership"
                 uib-btn-radio="'delegated'">
                 Delegated Docs
-              <span>({{(list |
-                filterByDue:filterParams.due:filterParams.dateRange |
-                filterByStatus:filterParams.documentStatus |
-                filterByOwnership:'delegated').length}})
+              <span>
+                ({{
+                  ((((list | filterByDueExpiry:filterParams.due:'activity_date_time':filterParams.dateRange)).length ?
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/views/dashboard/documents.html]: Same as above

File: uk.co.compucorp.civicrm.tasksassignments/views/dashboard/documents.html [hunk 5]
```diff
@@ -64,9 +72,14 @@ <h3 class="panel-title">{{label[filterParams.due]}}</h3>
                 ng-model="filterParams.ownership"
                 uib-btn-radio="null">
                 All
-              <span>({{(list |
-                filterByDue:filterParams.due:filterParams.dateRange |
-                filterByStatus:filterParams.documentStatus).length}})
+              <span>
+                ({{
+                  ((((list | filterByDueExpiry:filterParams.due:'activity_date_time':filterParams.dateRange)).length ?
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/views/dashboard/documents.html]: Same as above

File: uk.co.compucorp.civicrm.tasksassignments/views/dashboard/documents.html [hunk 6]
```diff
@@ -140,8 +153,9 @@ <h3 class="panel-title">{{label[filterParams.due]}}</h3>
               <tbody>
                 <tr
                   ng-repeat="document in
-                    listPaginated = (listFiltered = (listOngoing = (list)
-                    | filterByDue:filterParams.due:filterParams.dateRange
+                    listPaginated = (listFiltered = (listOngoing = (((list | filterByDueExpiry:filterParams.due:'activity_date_time':filterParams.dateRange)).length ?
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/views/dashboard/documents.html]: Same as above

File: uk.co.compucorp.civicrm.tasksassignments/views/directives/sidebar-filters.html [hunk 7]
```diff
@@ -5,13 +5,15 @@
         {{label[filter.value]}}
         <i class="fa fa-calendar" ng-click="viewInCalendar(filter.calendarView)"></i>
         <span class="badge badge-{{filter.badgeClass}}">
-          {{(list
+          {{(
+            (((list | filterByDueExpiry:filter.value:'activity_date_time':filterParams.dateRange)).length ?
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/views/directives/sidebar-filters.html]: Same as above

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/filters/due-expiry.filter.js (hunk 2 above)
Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/filters/due-expiry.filter.js]: No need, removed. Thanks!

File: uk.co.compucorp.civicrm.tasksassignments/views/dashboard/documents.html (hunk 3 above)
Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/views/dashboard/documents.html]: Moved to a function in controller.  Done!

File: uk.co.compucorp.civicrm.tasksassignments/views/dashboard/documents.html (hunk 4 above)
Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/views/dashboard/documents.html]: done!

File: uk.co.compucorp.civicrm.tasksassignments/views/dashboard/documents.html (hunk 5 above)
Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/views/dashboard/documents.html]: done!

File: uk.co.compucorp.civicrm.tasksassignments/views/dashboard/documents.html (hunk 6 above)
Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/views/dashboard/documents.html]: done!

File: uk.co.compucorp.civicrm.tasksassignments/views/directives/sidebar-filters.html (hunk 7 above)
Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/views/directives/sidebar-filters.html]: done!

**Comment by igorpavlov** [comment]
PR description:
- typo `the 'document` - should not be `'`
-`the documents` should be just `documents` because we are not talking about particular documents
- `Renamed the filter name as filterByDueExpiry` should be `Renamed the filter name to filterByDueExpiry`
- `Add afunction whic ` :)

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/task-list.js [hunk 8]
```diff
@@ -135,6 +135,15 @@ define([
             };
 
             /**
+             * Filters the tasks list based on filter type and due or expiry date
+             * @param {string} type filter type
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/task-list.js]: Please add another line with `*` between the function description and params.

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/tasks-list.spec.js [hunk 9]
```diff
@@ -82,6 +83,40 @@ define([
           });
         });
 
+        describe('filterByDateField()', function () {
+          var filteredTaskList;
+
+          describe('filtering tasks in between 2017-04-10 and 2017-04-20', function () {
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/tasks-list.spec.js]: I am not quite understand why `2017-04-10 - 2017-04-20` returns 3 documents and `2017-04-23 and 2017-04-25` returns 0. May be it would be better to describe tests like that:

describe: 'filtering tasks for the time period before due date'
it: 'returns filtered tasks by due date'

describe: 'filtering tasks for the time period after due date'
it: 'does not return filtered tasks'

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js [hunk 10]
```diff
@@ -180,7 +180,7 @@ define([
         });
 
         it('returns filtered document by expiry date of the document', function () {
-          expect(filteredDocumentList.length).toBe(2);
+          expect(filteredDocumentList.length).toBe(3);
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js]: Why so? Mocks have been changed?

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/task-list.js [hunk 11]
```diff
@@ -1,376 +1,383 @@
-define([
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/task-list.js]: Did you simply run semistandard * --fix ?
Any manual changes?

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/tasks-list.spec.js (hunk 9 above)
Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/tasks-list.spec.js]: Will update it!

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/task-list.js (hunk 11 above)
Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/task-list.js]: Well, its because I ran `run semistandard * --fix`. With some manual changes.

Comment (igorpavlov) [reply uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/task-list.js]: Could you point out these manual changes? What were they: indentation, names, logic?

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js (hunk 10 above)
Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js]: No, the the previous test case was not correct.

Comment (igorpavlov) [reply uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js]: Did it pass before? Why it has been changed from 2 to 3?

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/task-list.js (hunk 11 above)
Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/task-list.js]: Line no: 4, 11, 268,  are changes from `==` to `===`. And line no 357-366 is small logic separation as there was mixing of many `&&` and `||`.
